src/contract/consensus-data.ts

```diff
@@ -9,7 +9,53 @@ export interface ConsensusData {
 export function encodeConsensusData(consensusData: ConsensusData): string {
   const proposedOutcome = encodeOutcome(consensusData.proposedOutcome);
   return defaultAbiCoder.encode(
-    ['tuple(uint32, bytes)'],
+    ['tuple(uint32 furtherVotesRequired, bytes proposedOutcome)'],
     [[consensusData.furtherVotesRequired, proposedOutcome]],
   );
 }
+
+export function decodeConsensusData(appData: string): ConsensusData {
+  const {furtherVotesRequired, proposedOutcome} = defaultAbiCoder.decode(
+    ['tuple(uint32 furtherVotesRequired, bytes proposedOutcome)'],
+    appData,
+  )[0];
+  return {furtherVotesRequired, proposedOutcome: decodeOutcome(proposedOutcome)};
+}
+
+export function veto(
+  previousConsensusData: ConsensusData,
+  numberOfParticipants: number,
+): ConsensusData {
+  return propose(previousConsensusData, numberOfParticipants);
+}
+
+export function propose(consensusData: ConsensusData, numberOfParticipants: number): ConsensusData {
+  return {
+    proposedOutcome: consensusData.proposedOutcome,
+    furtherVotesRequired: numberOfParticipants - 1,
+  };
+}
+
+export function vote(consensusData: ConsensusData): ConsensusData {
+  return {
+    proposedOutcome: consensusData.proposedOutcome,
+    furtherVotesRequired: consensusData.furtherVotesRequired - 1,
+  };
+}
+
+export function finalVote(consensusData: ConsensusData): ConsensusData {
+  if (consensusData.furtherVotesRequired !== 0) {
+    throw new Error(
+      `Expected furtherVotesRequired to be 1, received ${consensusData.furtherVotesRequired} instead.`,
+    );
+  }
+  // TODO: if it's the final vote we return zeroed out consensusData
```

Comment:
Is this TODO still needed?

Reply:
No, I think this should just be a comment clarifying what we're doing.